range-requests/src/lib.rs

```diff
@@ -0,0 +1,441 @@
+    content_type: Option<&str>,
+) -> hyper::http::response::Builder {
+    let mut res = Response::builder();
+    res = res.header(ACCEPT_RANGES, "bytes");
+    res = res.header(
+        CONTENT_TYPE,
+        content_type.unwrap_or("application/octet-stream"),
+    );
+
+    if let Some(range) = range {
+        res = res.header(CONTENT_LENGTH, range.content_length().to_string());
+        res = res.header(CONTENT_RANGE, range.to_content_range());
+        res = res.status(StatusCode::PARTIAL_CONTENT);
+    } else {
+        res = res.header(CONTENT_LENGTH, file_length.to_string());
+        res = res.status(StatusCode::OK);
+    }
+
+    res
+}
+
+/// Represents the raw, unparsed values of "range" from a request header.
+pub struct PotentialRange(Vec<u8>);
+
+impl PotentialRange {
+    /// Parses a single range request out of the range request.
+    ///
+    /// `len` is the total length of the document, for the range request being made.
+    ///
+    /// On failure, returns a range response with the appropriate headers
+    /// to inform the caller how to make a correct range request.
+    pub fn parse(&self, len: u64) -> Result<SingleRange, Response<Body>> {
+        self.single_range(len).map_err(|_err| bad_range_response(len))
+    }
+
+    fn single_range(&self, len: u64) -> Result<SingleRange, Error> {
+        match http_range::HttpRange::parse_bytes(&self.0, len) {
+            Ok(ranges) => {
+                if ranges.len() != 1 || ranges[0].length < 1 {
+                    // Right now, we don't want to deal with encoding a
+                    // response that has multiple ranges.
+                    Err(Error::MultipleRangesUnsupported)
+                } else {
+                    Ok(SingleRange::new(ranges[0], len)?)
+                }
+            }
+            Err(err) => Err(Error::Parse(err)),
+        }
+    }
+}
+
+/// A parsed range request, and associated "total document length".
+#[derive(Clone, Debug)]
+pub struct SingleRange {
+    range: http_range::HttpRange,
+    total: u64,
+}
+
+#[cfg(test)]
+impl PartialEq for SingleRange {
+    fn eq(&self, other: &Self) -> bool {
+        self.range.start == other.range.start
+            && self.range.length == other.range.length
+            && self.total == other.total
+    }
+}
+
+impl SingleRange {
+    pub fn new(
+        range: http_range::HttpRange,
+        total: u64,
+    ) -> Result<Self, Error> {
+        let http_range::HttpRange { start, mut length } = range;
+
+        const INVALID_RANGE: Error =
+            Error::Parse(http_range::HttpRangeParseError::InvalidRange);
+
+        // Clip the length to avoid going beyond the end of the total range
+        if start.checked_add(length).ok_or(INVALID_RANGE)? >= total {
+            length = total.checked_sub(start).ok_or(INVALID_RANGE)?;
+        }
+        // If the length is zero, we cannot satisfy the range request
+        if length == 0 {
+            return Err(INVALID_RANGE);
+        }
```

Comment:
I'm not sure if reusing `HttpRangeParseError::InvalidRange` here is the right move. In this case, these error conditions indicate ranges that are syntactially valid (e.g. we parsed them successfully), but don't make sense. Per MDN, [such cases should return a 419 Range Not Satisfiable error](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/416). Looking at the `http_range` crate source, it looks like `InvalidRange` is returned when a range cannot be parsed (i.e. it's _syntactically_ invalid). See:
- https://docs.rs/http-range/latest/src/http_range/lib.rs.html#41-43
- https://docs.rs/http-range/latest/src/http_range/lib.rs.html#78-99
- https://docs.rs/http-range/latest/src/http_range/lib.rs.html#114-132
and so forth.

I think we might be better off defining our own error variant for parse-able but nonsensical ranges, and having that error variant return a 419 status code in the `From<Error> for HttpError` impl.

Reply:
I'll make these explicit error cases - though everything in this bucket is getting lumped into 416 anyway, which I think is fine? At least it makes tests more clear.